### PR TITLE
New Board: Rock 4C+ (rock-4c-plus)

### PR DESCRIPTION
**Pull request by tactmaster**
Raspberry Pi 4 form factor board 

https://radxa.com/products/rock4/4cp/

### Spec summary
Rockchip RK3399‑T
Dual-Core ARM® Cortex-A72 @ 1.5GHz and Quad-Core ARM® Cortex-A53 @ 1.0GHz
 4GB LPDDR4
microSD or eMMC
2x Micro HDMI ports support displays up to 4Kp60 resolution and 2Kp60
1x Gigabit Ethernet 
1x USB3 OTG/HOST port
1x USB3 HOST port
2x USB2 HOST ports
Raspberry Pi compatible GPIO Header

## Notes
https://wiki.radxa.com/Rock4/dev
Custom sdcard layout needed:
[radxa uboot docs ](https://wiki.radxa.com/Rockpi4/dev/u-boot)
Ethernet kernel module:
CONFIG_DWMAC_ROCKCHIP
Also HDMI requires kernel modules, core-image-minimal contained  kernel modules board by default. 
IMAGE_INSTALL:append = " kernel-modules "

### CODE REVIEW

File: kas/include/rockchip.yml [hunk 1]
```diff
@@ -0,0 +1,25 @@
+header:
+  version: 14
+
+repos:
+  meta-arm:
```
Comment (TheYoctoJester) [inline kas/include/rockchip.yml]: Is there a special requirement concerning the revision of `meta-arm`? If not, please remove and use `include/arm.yml`.

File: kas/rock-4c-plus.yml [hunk 2]
```diff
@@ -0,0 +1,17 @@
+header:
+  version: 14
+  includes:
+  - kas/include/mender-full.yml
+  - kas/include/rockchip.yml
+
+machine: rock-4c-plus
+
+# This is the setup for emmc the rock-4c-plus has other has emmc for that the block device is /dev/mmcblk0
```
Comment (TheYoctoJester) [inline kas/rock-4c-plus.yml]: I can't follow this comment, sorry. Can you rephrase it?

File: meta-mender-rockchip/README.md [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+https://wiki.radxa.com/Rockpi4/dev/serial-console
+
+    baudrate: 1500000
+    data bit: 8
+    stop bit: 1
+    parity  : none
+    flow control: none
+
+UART being connect can stop it from booting
```
Comment (TheYoctoJester) [inline meta-mender-rockchip/README.md]:
```suggestion
UART being connected can stop it from booting
```

File: meta-mender-rockchip/classes/rockchip-sdimg.bbclass [hunk 4]
```diff
@@ -0,0 +1,15 @@
+inherit image_types
+
+# It is not possible to specific this using MENDER_IMAGE_BOOTLOADER_FILE/OFFSET,
```
Comment (TheYoctoJester) [inline meta-mender-rockchip/classes/rockchip-sdimg.bbclass]:
```suggestion
# It is not possible to specify this using MENDER_IMAGE_BOOTLOADER_FILE/OFFSET,
```

File: meta-mender-rockchip/classes/rockchip-sdimg.bbclass [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+# instead we use an append to "sdimg" to embedd the bootloader
```
Comment (TheYoctoJester) [inline meta-mender-rockchip/classes/rockchip-sdimg.bbclass]:
```suggestion
# instead we use an append to "sdimg" to embed the bootloader
```

File: kas/rock-4c-plus.yml [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+# kernel-module-dwmac-rk is the driver for the ethernet it is a kernel module (which can't be build in )
+# it isn't install in minimal yocto builds
+local_conf_header:
+  rock-4c-plus: |
+    MENDER_UBOOT_STORAGE_INTERFACE = "mmc"
+    MENDER_UBOOT_STORAGE_DEVICE = "1"
+    IMAGE_CLASSES += "rockchip-sdimg"
+    MENDER_STORAGE_DEVICE = "/dev/mmcblk1"
```
Comment (TheYoctoJester) [inline kas/rock-4c-plus.yml:18]: Missing newline

File: kas/include/rockchip.yml [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+    layers:
+      meta-mender-rockchip:
+
+local_conf_header:
+  rockchip: |
+    MENDER_FEATURES_ENABLE:append = " mender-uboot mender-image mender-image-sd"
+    MENDER_FEATURES_DISABLE:append = " mender-grub mender-image-uefi"
+    IMAGE_FSTYPES:remove = "wic wic.bmap mender.bmap"
+    MENDER_PARTITION_ALIGNMENT = "10485760"
+    IMAGE_INSTALL:append = " kernel-modules "
```
Comment (TheYoctoJester) [inline kas/include/rockchip.yml]: Missing newline

File: meta-mender-rockchip/templates/local.conf.append
Comment (TheYoctoJester) [inline meta-mender-rockchip/templates/local.conf.append:1]: Do we actually need this file? The information should be already provided through the `kas` file

File: meta-mender-rockchip/templates/bblayers.conf.sample
Comment (TheYoctoJester) [inline meta-mender-rockchip/templates/bblayers.conf.sample:1]: Do we actually need this file? The information should be already provided through the kas file

File: meta-mender-rockchip/recipes-bsp/u-boot/u-boot_%.bbappend [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+
+
+# Add uboot envirment offset that didn't get by mender autoconfigure
+# Will boot but fw_printenv do not work.
+
+SRC_URI += "file://0001-Add-mender-settings.patch    \
+"
+MENDER_UBOOT_AUTO_CONFIGURE = "1"
+MENDER_UBOOT_ENV_STORAGE_DEVICE_OFFSET_1 = "0xa00000"
+MENDER_UBOOT_ENV_STORAGE_DEVICE_OFFSET_2 = "0x1400000"
```
Comment (TheYoctoJester) [inline meta-mender-rockchip/recipes-bsp/u-boot/u-boot_%.bbappend]: Missing newline

File: meta-mender-rockchip/recipes-bsp/u-boot/u-boot_%.bbappend [hunk 9]
```diff
@@ -0,0 +1,13 @@
+
```
Comment (TheYoctoJester) [inline meta-mender-rockchip/recipes-bsp/u-boot/u-boot_%.bbappend]: Unnecessary blank lines.

**Review by TheYoctoJester — CHANGES_REQUESTED** [review]
Thanks for the submission, that's really a great contribution. I have added a couple of somewhat minor issues, of which none is really critical. If you add me (@TheYoctoJester) as contributor, I can also address those.

The patch history also needs some clean up, will address that separately.

Oh, and all patches need a "Signed-off-by:" 👍

File: meta-mender-rockchip/recipes-bsp/u-boot/u-boot/0001-qemu-arm-make-QFW-MMIO-implied-on-qemu-arm.patch
Comment (TheYoctoJester) [inline meta-mender-rockchip/recipes-bsp/u-boot/u-boot/0001-qemu-arm-make-QFW-MMIO-implied-on-qemu-arm.patch:1]: This commit should be merged with the previous one.

File: meta-mender-rockchip/recipes-bsp/u-boot/u-boot_%.bbappend
Comment (TheYoctoJester) [inline meta-mender-rockchip/recipes-bsp/u-boot/u-boot_%.bbappend:1]: This commit should be merged into the previous one.

**Comment by TheYoctoJester** [comment]
@tactmaster are you okay if I take the liberty to rebase on top of current `scarthgap`, or do you want to do it?

File: meta-mender-rockchip/templates/bblayers.conf.sample
Comment (tactmaster) [reply meta-mender-rockchip/templates/bblayers.conf.sample:1]: No you are right it just duplicate information

File: kas/include/rockchip.yml (hunk 1 above)
Comment (tactmaster) [reply kas/include/rockchip.yml]: Ok I will test that.

**Comment by tactmaster** [comment]
Forced pushs was to add signin-off. No code changes https://github.com/mendersoftware/meta-mender-community/compare/93c29bd91d986fadfa5177c0decc52fb7471a0a0..fc6a11cab0cc3c858a2441288b91917446b87cac

**Comment by tactmaster** [comment]
> @tactmaster are you okay if I take the liberty to rebase on top of current `scarthgap`, or do you want to do it?

I have rebased. No see any problem with that

**Comment by TheYoctoJester** [comment]
> > @tactmaster are you okay if I take the liberty to rebase on top of current `scarthgap`, or do you want to do it?
> 
> I have rebased. No see any problem with that

c046e61188d4d340fe5700f6084b162528d1cca3 and f545e5323edbaf0935e0ae59720de8aa011e8aa7 still lack the sign-off.

If you add me as contributor, I can fix the remaining things.

File: kas/include/rockchip.yml (hunk 1 above)
Comment (tactmaster) [reply kas/include/rockchip.yml]: Ok I checked it is a requirement of meta-rockchip  :-)
https://git.yoctoproject.org/meta-rockchip/tree/conf/layer.conf

File: kas/rock-4c-plus.yml (hunk 2 above)
Comment (tactmaster) [reply kas/rock-4c-plus.yml]: Sorry miss type there
The setup below is for SD card, for emmc the block device is /dev/mmcblk0

**Comment by tactmaster** [comment]
@TheYoctoJester 
Thanks for you quick reply.
I will get the signoff fix in a few days time. Also i think I may experiment  with a bitbake linter :)
I won't add you a contributor just yet for internal reasons.

File: kas/include/rockchip.yml (hunk 1 above)
Comment (TheYoctoJester) [reply kas/include/rockchip.yml]: > Ok I checked it is a requirement of meta-rockchip :-) https://git.yoctoproject.org/meta-rockchip/tree/conf/layer.conf

No, I meant the specific revision. There is already an include for `meta-arm` at `kas/include/arm.yml`, and unless necessary I would like to avoid duplication.

**Comment by tactmaster** [comment]
@TheYoctoJester  OK I have hopefully signed off all commit and I also did some tiding on formatting etc. 
Thanks :)

**Comment by TheYoctoJester** [comment]
Hi @tactmaster, thanks! Did some additional clean up and a build test - to be sure I didn't break anything, can you give this a last functional test? Upon your 👍 I will merge then.

**Review by TheYoctoJester — APPROVED** [review]

**Comment by tactmaster** [comment]
@TheYoctoJester I started a fresh build folder, fresh burn to sdcard.  tested an upgrade (via mender server) and it all worked fine. Change from parition from 2 to 3 as root 
Also I assumed your phone number is in git.txt https://xkcd.com/1597/ :)

**Comment by TheYoctoJester** [comment]
> Also I assumed your phone number is in git.txt https://xkcd.com/1597/ :)

Not yet ;-)
